LocalFinalVariableName: false negative on pattern variable

https://checkstyle.org/config_naming.html#LocalFinalVariableName

````
$ cat TestClass.java
public class Test {
    public void run(String... arguments) {
        final Object o = 45;
        if (o instanceof final String BAD) {
            System.out.println(BAD);
        }
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="LocalFinalVariableName"/>
<module name="LocalVariableName"/>
    </module>
</module>

$ java -jar checkstyle-9.0.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

I am expecting a violation on line 5 since `BAD` is a local, final variable and it doesn't start with a lower case as is default required.

```
public class Test {
    public void run(String... arguments) {
        final Object o = 45;
        if (o instanceof final String BAD) {
            o = 34;  // javac: Test.java:5: error: cannot assign a value to final variable o
            System.out.println(BAD);
        }
    }
}
```

For usual local variables, we have `LocalVariableName` and ` LocalFinalVariableName`. `LocalVariableName` does not check variables declared as `final`.

If we want to check pattern variables declared as `final` differently, then perhaps analogous idea with pattern variables should hold:

```
public class Test {
    public void run(String... arguments) {
        final Object BAD1 = 45; // violation by LocalFinalVariableName only
        Object BAD2 = 42;  // violation by LocalVariableName only
        Object o = 42;
        if (o instanceof final String BAD3) { // should be FinalPatternVariableName?
            System.out.println(BAD3);
        }
        if (o instanceof String BAD4) { // should be PatternVariableName only
            System.out.println(BAD4);
        }
    }
}

```

Current behavior:

```
➜  company cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="LocalFinalVariableName"/>
        <module name="LocalVariableName"/>
        <module name="PatternVariableName"/>
    </module>
</module>  

➜  java -jar checkstyle-9.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:3:22: Name 'BAD1' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LocalFinalVariableName]
[ERROR] Test.java:4:16: Name 'BAD2' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LocalVariableName]
[ERROR] Test.java:6:39: Name 'BAD3' must match pattern '^[a-z][a-zA-Z0-9]*$'. [PatternVariableName]
[ERROR] Test.java:9:33: Name 'BAD4' must match pattern '^[a-z][a-zA-Z0-9]*$'. [PatternVariableName]
Audit done.
Checkstyle ends with 4 errors.
```

So, any fix for this issue should also exclude final pattern variables from being checked in `PatternVariableNameCheck`.

Here is another idea: now that the statement from [here](https://docs.oracle.com/en/java/javase/14/docs/specs/patterns-instanceof-jls.html#jls-6.3.1) "Pattern variables are implicitly final by design" is not accurate any longer, would it make sense to just add pattern variable support to `LocalVariableName` and `LocalFinalVariableName` and do away with `PatternVariableName`?

I think the two best options are either:
(1) Make a new check, `FinalPatternVariableNameCheck`, or
(2) Add support to `LocalVariableName` and `LocalFinalVariableName` and do away with `PatternVariableName`

Pattern matching/ variable extension is imminent, as seen in [JEP406](https://openjdk.java.net/jeps/406) and [JEP405](https://openjdk.java.net/jeps/405), we may regret grouping pattern variables in with usual local variables later.

We don't need a new check for this, we can easily achieve this behavior with PatternVariableCheck and a SuppressionXpathSingleFilter module:

```
➜  src cat Test.java 
public class Test {

    void m1(Object o) {
        // ok below, not final
        if (o instanceof String BAD) {
            System.out.println(BAD);
        }
    }

    void m2(Object o) {
        // violation below, final
        if (o instanceof final String BAD) {
            System.out.println(BAD);
        }
    }
}

➜  src cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="PatternVariableName"/>
    <module name="SuppressionXpathSingleFilter">
      <property name="checks" value="PatternVariableName"/>
      <property name="query" value="//PATTERN_VARIABLE_DEF[not(./MODIFIERS/FINAL)]/IDENT"/>
    </module>
  </module>
</module>

➜  src javac Test.java
                                             
➜  src java -jar checkstyle-10.13.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:12:39: Name 'BAD' must match pattern '^[a-z][a-zA-Z0-9]*$'. [PatternVariableName]
Audit done.
Checkstyle ends with 1 errors.
```

@rnveach if you agree, let's close this issue

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalFinalVariableNameCheck.java
@@ -30,6 +30,11 @@ import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
  *  A catch parameter and resources in try statements
  * are considered to be a local, final variables.
  * </p>
+ * <p>
+ * This check does not support final pattern variables. Instead, use
+ * <a href="https://checkstyle.org/checks/naming/patternvariablename.html#PatternVariableName">
+ * PatternVariableName</a>.
+ * </p>
  * <ul>
  * <li>
  * Property {@code format} - Sets the pattern to match valid identifiers.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheck.java
@@ -29,6 +29,11 @@ import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
  * A catch parameter is considered to be
  * a local variable.
  * </p>
+ * <p>
+ * This check does not support pattern variables. Instead, use
+ * <a href="https://checkstyle.org/checks/naming/patternvariablename.html#PatternVariableName">
+ * PatternVariableName</a>.
+ * </p>
  * <ul>
  * <li>
  * Property {@code allowOneCharVarInForLoop} - Allow one character variable name in

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/LocalFinalVariableNameCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/LocalFinalVariableNameCheck.xml
@@ -8,6 +8,11 @@
  Checks that local final variable names conform to a specified pattern.
   A catch parameter and resources in try statements
  are considered to be a local, final variables.
+ &lt;/p&gt;
+ &lt;p&gt;
+ This check does not support final pattern variables. Instead, use
+ &lt;a href="https://checkstyle.org/checks/naming/patternvariablename.html#PatternVariableName"&gt;
+ PatternVariableName&lt;/a&gt;.
  &lt;/p&gt;</description>
          <properties>
             <property default-value="^([a-z][a-zA-Z0-9]*|_)$"

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/LocalVariableNameCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/LocalVariableNameCheck.xml
@@ -8,6 +8,11 @@
  Checks that local, non-{@code final} variable names conform to a specified pattern.
  A catch parameter is considered to be
  a local variable.
+ &lt;/p&gt;
+ &lt;p&gt;
+ This check does not support pattern variables. Instead, use
+ &lt;a href="https://checkstyle.org/checks/naming/patternvariablename.html#PatternVariableName"&gt;
+ PatternVariableName&lt;/a&gt;.
  &lt;/p&gt;</description>
          <properties>
             <property default-value="false"

--- a/src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java
@@ -406,6 +406,8 @@ public final class InlineConfigParser {
                 "com.puppycrawl.tools.checkstyle.checks.naming.ParameterNameCheck");
         moduleMappings.put("MethodName",
                 "com.puppycrawl.tools.checkstyle.checks.naming.MethodNameCheck");
+        moduleMappings.put("SuppressionXpathSingleFilter",
+                "com.puppycrawl.tools.checkstyle.filters.SuppressionXpathSingleFilter");
 
         String fullyQualifiedClassName;
         if (moduleMappings.containsKey(moduleName)) {

--- a/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheckExamplesTest.java
+++ b/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheckExamplesTest.java
@@ -68,4 +68,15 @@ public class PatternVariableNameCheckExamplesTest extends AbstractExamplesModule
 
         verifyWithInlineConfigParser(getNonCompilablePath("Example3.java"), expected);
     }
+
+    @Test
+    public void testExample4() throws Exception {
+
+        final String[] expected = {
+            "37:30: " + getCheckMessage(MSG_INVALID_PATTERN, "BAD", "^([a-z][a-zA-Z0-9]*|_)$"),
+            "40:36: " + getCheckMessage(MSG_INVALID_PATTERN, "bad", "^[A-Z][A-Z0-9]*$"),
+        };
+
+        verifyWithInlineConfigParser(getNonCompilablePath("Example4.java"), expected);
+    }
 }

--- a/src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/Example4.java
+++ b/src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/Example4.java
@@ -1,0 +1,43 @@
+/*xml
+<module name="Checker">
+  <module name="TreeWalker">
+    <module name="PatternVariableName">
+        <property name="id" value="FinalPatternVariableName"/>
+        <property name="format" value="^[A-Z][A-Z0-9]*$"/>
+    </module>
+
+    <module name="PatternVariableName">
+        <property name="id" value="NonFinalPatternVariableName"/>
+        <property name="format" value="^([a-z][a-zA-Z0-9]*|_)$"/>
+    </module>
+
+    <module name="SuppressionXpathSingleFilter">
+      <property name="id" value="FinalPatternVariableName"/>
+      <property name="query" value="//PATTERN_VARIABLE_DEF[
+            not(./MODIFIERS/FINAL)]/IDENT"/>
+    </module>
+
+    <module name="SuppressionXpathSingleFilter">
+      <property name="id" value="NonFinalPatternVariableName"/>
+      <property name="query" value="//PATTERN_VARIABLE_DEF[
+            (./MODIFIERS/FINAL)]/IDENT"/>
+    </module>
+  </module>
+</module>
+
+
+*/
+
+//non-compiled with javac: Compilable with Java17
+package com.puppycrawl.tools.checkstyle.checks.naming.patternvariablename;
+
+// xdoc section -- start
+class Example4 {
+  void foo(Object o1){
+    if (o1 instanceof String BAD) {} // violation
+    if (o1 instanceof String good) {}
+    if (o1 instanceof final String GOOD) {}
+    if (o1 instanceof final String bad) {} // violation
+  }
+}
+// xdoc section -- end

--- a/src/xdocs/checks/naming/localfinalvariablename.xml
+++ b/src/xdocs/checks/naming/localfinalvariablename.xml
@@ -14,6 +14,11 @@
           A catch parameter and resources in try statements
           are considered to be a local, final variables.
         </p>
+        <p>
+          This check does not support final pattern variables. Instead, use
+          <a href="patternvariablename.html#PatternVariableName">
+          PatternVariableName</a>.
+        </p>
       </subsection>
 
       <subsection name="Properties" id="Properties">

--- a/src/xdocs/checks/naming/localfinalvariablename.xml.template
+++ b/src/xdocs/checks/naming/localfinalvariablename.xml.template
@@ -14,6 +14,11 @@
           A catch parameter and resources in try statements
           are considered to be a local, final variables.
         </p>
+        <p>
+          This check does not support final pattern variables. Instead, use
+          <a href="patternvariablename.html#PatternVariableName">
+          PatternVariableName</a>.
+        </p>
       </subsection>
 
       <subsection name="Properties" id="Properties">

--- a/src/xdocs/checks/naming/localvariablename.xml
+++ b/src/xdocs/checks/naming/localvariablename.xml
@@ -13,6 +13,11 @@
           Checks that local, non-<code>final</code> variable names conform to a specified pattern.
           A catch parameter is considered to be a local variable.
         </p>
+        <p>
+          This check does not support pattern variables. Instead, use
+          <a href="patternvariablename.html#PatternVariableName">
+          PatternVariableName</a>.
+        </p>
       </subsection>
 
       <subsection name="Properties" id="Properties">

--- a/src/xdocs/checks/naming/localvariablename.xml.template
+++ b/src/xdocs/checks/naming/localvariablename.xml.template
@@ -13,6 +13,11 @@
           Checks that local, non-<code>final</code> variable names conform to a specified pattern.
           A catch parameter is considered to be a local variable.
         </p>
+        <p>
+          This check does not support pattern variables. Instead, use
+          <a href="patternvariablename.html#PatternVariableName">
+          PatternVariableName</a>.
+        </p>
       </subsection>
 
       <subsection name="Properties" id="Properties">

--- a/src/xdocs/checks/naming/patternvariablename.xml
+++ b/src/xdocs/checks/naming/patternvariablename.xml
@@ -106,6 +106,48 @@ class Example3 {
   }
 }
         </source>
+        <p id="Example4-config">
+          An example of how to configure the check with different format for final and non-final
+          pattern variables:
+        </p>
+        <source>
+&lt;module name=&quot;Checker&quot;&gt;
+  &lt;module name=&quot;TreeWalker&quot;&gt;
+    &lt;module name=&quot;PatternVariableName&quot;&gt;
+        &lt;property name=&quot;id&quot; value=&quot;FinalPatternVariableName&quot;/&gt;
+        &lt;property name=&quot;format&quot; value=&quot;^[A-Z][A-Z0-9]*$&quot;/&gt;
+    &lt;/module&gt;
+
+    &lt;module name=&quot;PatternVariableName&quot;&gt;
+        &lt;property name=&quot;id&quot; value=&quot;NonFinalPatternVariableName&quot;/&gt;
+        &lt;property name=&quot;format&quot; value=&quot;^([a-z][a-zA-Z0-9]*|_)$&quot;/&gt;
+    &lt;/module&gt;
+
+    &lt;module name=&quot;SuppressionXpathSingleFilter&quot;&gt;
+      &lt;property name=&quot;id&quot; value=&quot;FinalPatternVariableName&quot;/&gt;
+      &lt;property name=&quot;query&quot; value=&quot;//PATTERN_VARIABLE_DEF[
+            not(./MODIFIERS/FINAL)]/IDENT&quot;/&gt;
+    &lt;/module&gt;
+
+    &lt;module name=&quot;SuppressionXpathSingleFilter&quot;&gt;
+      &lt;property name=&quot;id&quot; value=&quot;NonFinalPatternVariableName&quot;/&gt;
+      &lt;property name=&quot;query&quot; value=&quot;//PATTERN_VARIABLE_DEF[
+            (./MODIFIERS/FINAL)]/IDENT&quot;/&gt;
+    &lt;/module&gt;
+  &lt;/module&gt;
+&lt;/module&gt;
+        </source>
+        <p id="Example4-code">Code Example:</p>
+        <source>
+class Example4 {
+  void foo(Object o1){
+    if (o1 instanceof String BAD) {} // violation
+    if (o1 instanceof String good) {}
+    if (o1 instanceof final String GOOD) {}
+    if (o1 instanceof final String bad) {} // violation
+  }
+}
+        </source>
       </subsection>
 
       <subsection name="Example of Usage" id="Example_of_Usage">

--- a/src/xdocs/checks/naming/patternvariablename.xml.template
+++ b/src/xdocs/checks/naming/patternvariablename.xml.template
@@ -69,6 +69,21 @@
                  value="resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/Example3.java"/>
           <param name="type" value="code"/>
         </macro>
+        <p id="Example4-config">
+          An example of how to configure the check with different format for final and non-final
+          pattern variables:
+        </p>
+        <macro name="example">
+          <param name="path"
+                 value="resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/Example4.java"/>
+          <param name="type" value="config"/>
+        </macro>
+        <p id="Example4-code">Code Example:</p>
+        <macro name="example">
+          <param name="path"
+                 value="resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/patternvariablename/Example4.java"/>
+          <param name="type" value="code"/>
+        </macro>
       </subsection>
 
       <subsection name="Example of Usage" id="Example_of_Usage">

